### PR TITLE
lgc/dialect: add buffer.length and buffer.ptr.diff

### PR DESCRIPTION
**Pull request by nhaehnle**
Builds on #2023.

Convert existing "LGC intrinsics" to start a "sub-dialect" for buffer
fat pointers, i.e. LLVM IR pointers whose underlying representation is
a buffer descriptor -- essentially a base + buffer size -- and an
offset into the buffer.

buffer.desc.to.ptr is still missing because of opaque pointer troubles.
Basically, llvm-dialects doesn't know how to handle an operation that
returns a non-opaque pointer, and it doesn't make much sense to spend
effort on adding that support anymore. So I plan to add
buffer.desc.to.ptr once we've fully transitioned to opaque pointers.
    
Some `ashr / sdiv` become `lshr` in the test cases. This is okay because
the resulting value is truncated.

### CODE REVIEW

File: llpc/translator/lib/SPIRV/SPIRVReader.cpp
Comment (trenouf) [inline llpc/translator/lib/SPIRV/SPIRVReader.cpp:2832]: Further to comments on #2023, I see you are starting to use a dialect to replace and bypass the Builder interface. I misunderstood earlier discussion; I thought you were going to start with purely internal LGC stuff.

The question is: do we want to lose the "builder direct" path like this? The idea is that that would have saved a bit of compilation time, at the expense of not giving you semi-stable IR at the input-to-lgc level. But:

1. We have never used it.
2. I guess the dialect thing has potential to be less wasteful than the "builder recorder and replay" thing in that you don't need to lower everything in a replay phase as soon as you get in to LGC, only to need to lower some of it again later on.

So is everyone happy for this fundamental change to the lgc interface to start happening here?

File: lgc/patch/PatchBufferOp.cpp
Comment (Flakebi) [inline lgc/patch/PatchBufferOp.cpp:571]: Why was the else removed?

**Review by Flakebi — COMMENTED** [review]
One nit, otherwise LGTM

File: lgc/patch/PatchBufferOp.cpp
Comment (nhaehnle) [reply lgc/patch/PatchBufferOp.cpp:571]: The old code made an assumption about the prefix of the names of the "intrinsics". The new code no longer does that.

By the way, this probably makes the new version slightly less efficient under the current llvm-dialects implementation using string comparisons. My intention is to implement the same metadata "opcode" trick that the builder recorder uses and make it switchable so that we can actually measure the performance impact.

Comment (Flakebi) [reply lgc/patch/PatchBufferOp.cpp:571]: Makes sense. Good luck measuring the performance impact ;)

**Comment by amdvlk-admin** [comment]
### Test summary for commit f2c079b74863b686dbc2bb01de5731ec0ada55ad
<details><summary>CTS tests (Failed: 0/207274)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Ubuntu 22.04, Navi2x</summary><ul>

<li>Passed: 38905/70473 (55.2%)</li>
<li>Failed: 0/70473 (0.0%)</li>
<li>Not Supported: 31568/70473 (44.8%)</li><li>Warnings: 0/70473 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 35750/70481 (50.7%)</li>
<li>Failed: 0/70481 (0.0%)</li>
<li>Not Supported: 34731/70481 (49.3%)</li><li>Warnings: 0/70481 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 9.0, Gfx9</summary><ul>

<li>Passed: 33476/66320 (50.5%)</li>
<li>Failed: 0/66320 (0.0%)</li>
<li>Not Supported: 32844/66320 (49.5%)</li><li>Warnings: 0/66320 (0.0%)</li>
</ul></details>


</ul></details>

File: llpc/translator/lib/SPIRV/SPIRVReader.cpp
Comment (trenouf) [reply llpc/translator/lib/SPIRV/SPIRVReader.cpp]: I removed "builder direct" mode a while ago.

**Review by trenouf — APPROVED** [review]
LGTM (after rebase once the underneath commit is merged)

**Comment by amdvlk-admin** [comment]
### Test summary for commit ca304c9c0aa02b150b63b263056a5b05fb2cd1df
<details><summary>CTS tests (Failed: 0/282025)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 38862/70507 (55.1%)</li>
<li>Failed: 0/70507 (0.0%)</li>
<li>Not Supported: 31645/70507 (44.9%)</li><li>Warnings: 0/70507 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 22.04, Navi2x</summary><ul>

<li>Passed: 39112/70507 (55.5%)</li>
<li>Failed: 0/70507 (0.0%)</li>
<li>Not Supported: 31395/70507 (44.5%)</li><li>Warnings: 0/70507 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 35959/70509 (51.0%)</li>
<li>Failed: 0/70509 (0.0%)</li>
<li>Not Supported: 34550/70509 (49.0%)</li><li>Warnings: 0/70509 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 9.0, Gfx9</summary><ul>

<li>Passed: 35861/70502 (50.9%)</li>
<li>Failed: 0/70502 (0.0%)</li>
<li>Not Supported: 34641/70502 (49.1%)</li><li>Warnings: 0/70502 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by nhaehnle** [comment]
This now depends on an llvm-dialects update for the new `Memory<...>` trait functionality, changing to draft until https://github.com/GPUOpen-Drivers/llvm-dialects/pull/18 has landed.

**Comment by amdvlk-admin** [comment]
### Test summary for commit 0deb2828511660e196764d4e3ab35447ad286d3f
<details><summary>CTS tests (Failed: 0/282025)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 38862/70507 (55.1%)</li>
<li>Failed: 0/70507 (0.0%)</li>
<li>Not Supported: 31645/70507 (44.9%)</li><li>Warnings: 0/70507 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 22.04, Navi2x</summary><ul>

<li>Passed: 39112/70507 (55.5%)</li>
<li>Failed: 0/70507 (0.0%)</li>
<li>Not Supported: 31395/70507 (44.5%)</li><li>Warnings: 0/70507 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 35959/70509 (51.0%)</li>
<li>Failed: 0/70509 (0.0%)</li>
<li>Not Supported: 34550/70509 (49.0%)</li><li>Warnings: 0/70509 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 9.0, Gfx9</summary><ul>

<li>Passed: 35861/70502 (50.9%)</li>
<li>Failed: 0/70502 (0.0%)</li>
<li>Not Supported: 34641/70502 (49.1%)</li><li>Warnings: 0/70502 (0.0%)</li>
</ul></details>


</ul></details>

File: lgc/builder/BuilderBase.cpp
Comment (amdrexu) [inline lgc/builder/BuilderBase.cpp:50]: Better to remove "llvm::" prefix

File: lgc/builder/BuilderBase.cpp
Comment (amdrexu) [inline lgc/builder/BuilderBase.cpp:49]: I think for newly added function, we could start to add "@return : " comments for those have return types. This was once requested by Jakub as a coding style.

File: lgc/interface/lgc/LgcDialect.td
Comment (amdrexu) [inline lgc/interface/lgc/LgcDialect.td:40]: Not sure if this TODO could be enabled

**Comment by amdvlk-admin** [comment]
### Test summary for commit 9c3060203e7f154d248305fe65d50788e0462047
<details><summary>CTS tests (Failed: 0/282025)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 38862/70507 (55.1%)</li>
<li>Failed: 0/70507 (0.0%)</li>
<li>Not Supported: 31645/70507 (44.9%)</li><li>Warnings: 0/70507 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 22.04, Navi2x</summary><ul>

<li>Passed: 39112/70507 (55.5%)</li>
<li>Failed: 0/70507 (0.0%)</li>
<li>Not Supported: 31395/70507 (44.5%)</li><li>Warnings: 0/70507 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 35959/70509 (51.0%)</li>
<li>Failed: 0/70509 (0.0%)</li>
<li>Not Supported: 34550/70509 (49.0%)</li><li>Warnings: 0/70509 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 9.0, Gfx9</summary><ul>

<li>Passed: 35861/70502 (50.9%)</li>
<li>Failed: 0/70502 (0.0%)</li>
<li>Not Supported: 34641/70502 (49.1%)</li><li>Warnings: 0/70502 (0.0%)</li>
</ul></details>


</ul></details>

File: lgc/interface/lgc/LgcDialect.td
Comment (nhaehnle) [reply lgc/interface/lgc/LgcDialect.td:40]: It should be possible to do that soon, but I'd rather hold off for now because I believe the opaque pointer enabling patch hasn't landed yet in the internal branch?

**Comment by amdvlk-admin** [comment]
### Test summary for commit aa52a85fdc0a952fc3fae1f50cae99cbd6c973a8
<details><summary>CTS tests (Failed: 0/282025)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 38862/70507 (55.1%)</li>
<li>Failed: 0/70507 (0.0%)</li>
<li>Not Supported: 31645/70507 (44.9%)</li><li>Warnings: 0/70507 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 22.04, Navi2x</summary><ul>

<li>Passed: 39112/70507 (55.5%)</li>
<li>Failed: 0/70507 (0.0%)</li>
<li>Not Supported: 31395/70507 (44.5%)</li><li>Warnings: 0/70507 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 35959/70509 (51.0%)</li>
<li>Failed: 0/70509 (0.0%)</li>
<li>Not Supported: 34550/70509 (49.0%)</li><li>Warnings: 0/70509 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 9.0, Gfx9</summary><ul>

<li>Passed: 35861/70502 (50.9%)</li>
<li>Failed: 0/70502 (0.0%)</li>
<li>Not Supported: 34641/70502 (49.1%)</li><li>Warnings: 0/70502 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by nhaehnle** [comment]
ping @GPUOpen-Drivers/llpcapprovers

**Comment by amdvlk-admin** [comment]
### Test summary for commit e0b9b9aed8379d3921c7ab3484c0f0b349ad2f4c
<details><summary>CTS tests (Failed: 0/211487)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 39248/70495 (55.7%)</li>
<li>Failed: 0/70495 (0.0%)</li>
<li>Not Supported: 31247/70495 (44.3%)</li><li>Warnings: 0/70495 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 39498/70495 (56.0%)</li>
<li>Failed: 0/70495 (0.0%)</li>
<li>Not Supported: 30997/70495 (44.0%)</li><li>Warnings: 0/70495 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 9.0, Gfx10</summary><ul>

<li>Passed: 36345/70497 (51.6%)</li>
<li>Failed: 0/70497 (0.0%)</li>
<li>Not Supported: 34152/70497 (48.4%)</li><li>Warnings: 0/70497 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by nhaehnle** [comment]
ping @amdrexu

**Review by amdrexu — APPROVED** [review]

**Pull request merged**
